fix endpoint selection for wildcard to/fromEndpoint in CCNP

fristonio opened the pull request:
See individual commits for more details.

- [x] Fixes #12844 
- [x] Add unit tests.
- [x] Update pre-flight CCNP checks to check for possible required policy upgrade in cluster.

Example of preflight warning.
![image](https://user-images.githubusercontent.com/21292343/90757664-87b9f080-e2fb-11ea-959e-42476e024357.png)

```release-note
Fix endpoint selection for a wildcard to/fromEndpoints in CCNP.
Cilium will only allow access from Cilium-managed endpoints in such cases instead of allowing traffic from any source. Preflight checks, when following the upgrade guide, have been extended to warn users of the new behavior.
```

joestringer reviewed (commented):
I was expecting this change to be made in the same (or similar) place where we inject the namespace selector for CNP `EndpointSelector`s, but I don't see that logic here so it's hard to compare against that logic. Were you aware of that other code? Wondering if there's a functional tradeoff from having it here vs. wherever the namespace selector CNP injection logic is.

fristonio committed:
pkg/policy: fix endpoint selection for wildcard to/fromEndpoint in CCNP

* This commit fixes an issue in endpoint selection when we provide
  wildcard for to/fromEndpoint in CCNP. When a wildcard is provided in CCNP
  fromEndpoint selector we end up with a truly empty endpoint selector.
  This results in allowing all traffic. The commit restricts this to only
  include endpoints that are managed by cilium by checking the presence
  of namespace label in endpoint.

* For a more detailed explaination of the approach and the issue take a
  look at discussion following this github comment -
  https://github.com/cilium/cilium/pull/12890#issuecomment-674358012

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

joestringer commented on pkg/k8s/apis/cilium.io/utils/utils.go:
```diff
@@ -93,8 +93,17 @@ func getEndpointSelector(namespace string, labelSelector *slim_metav1.LabelSelec
-	if !matchesInit && !es.HasKey(podPrefixLbl) && !es.HasKey(podAnyPrefixLbl) && namespace != "" {
-		es.AddMatch(podPrefixLbl, namespace)
+	if !matchesInit && !es.HasKey(podPrefixLbl) && !es.HasKey(podAnyPrefixLbl) {
+		if namespace != "" {
```
nit: We will usually put the == case first so that when trying to interpret the `else` case we don't have to do the mental swap involved in  `! (namespace != "")`.

joestringer reviewed (changes requested):
Looks good to me, just one minor nit.

fristonio committed:
pkg/k8s: add unit tests for empty to/fromEndpoint with ccnp

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

fristonio committed:
cilium/preflight: add check for empty to/fromEndpoint in CCNP validation

* This commit extends the cilium preflight validate-cnp check. When
  validating CCNP it checks if there is an empty to/from endpoint
  selector in the rules and warns about the problem and a possible fix.

* This is to help users with upgrade scenarios when using Cilium. For a
  more detailed discussion on the probelm see issue -
  https://github.com/cilium/cilium/issues/12844

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+		resCCNP := cilium_v2.CiliumClusterwideNetworkPolicy{}
+		err = json.Unmarshal(ccnpBytes, &resCCNP)
+		if err != nil {
+			logger.Warnf("Error unmarshalling CCNP bytes: %s", err)
```
Please use structured logging here with `WithError()` instead.

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+			}
+		}
+	} else {
+		logger.Warnf("Error marshalling CCNP structure: %s", err)
```
Here as well

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+
+	// At this point we have validated the custom resource with the new CRV.
+	// We can try converting it to the new CCNP type.
+	// This should not fail so we are ignoring any errors here.
```
This comment says that we are ignoring errors, but there's a else-clause that logs the error. I wouldn't consider that ignoring :wink:

fristonio replied:
Ah, we are ignoring the error actually as we only log a warning instead of returning the error. But yeah, the comment is not clear. I will update it. :sweat_smile:

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+		for _, r := range rule.Ingress {
+			if len(r.FromEndpoints) > 0 {
+				for _, epSel := range r.FromEndpoints {
+					// Print a warning if the endpoint selector is a wildcard.
```
This comment is doesn't seem necessary anymore.

fristonio replied:
Oh, I forgot to remove this. :sweat_smile:

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator_test.go:
```diff
@@ -20,11 +20,16 @@ import (
 )
 
+var (
+	log = logrus.New()
```
I think we have a convention for instantiating the logger, see other packages.

christarazi replied:
(Actually, ignore this comment if superseded by the dependency-injection question above.)

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -18,21 +18,27 @@ import (
 	cnpValidator  *validate.SchemaValidator
 	ccnpValidator *validate.SchemaValidator
+
+	logger *logrus.Logger
```
Why is this necessary? Can we make it a package-level variable like how we do in other packages throughout the codebase? 

Removing this dependency-injection will also allow us to avoid instantiating the logger in the test code as well.

fristonio replied:
Yeah, this makes sense actually. We can also attach `validator` subsystem context to the logs. I will update the PR.

christarazi commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +109,90 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+		logfields.CiliumClusterwideNetworkPolicyName: ccnp.GetName(),
+	})
+
+	// At this point we have validated the custom resource with the new CRV.
+	// We can try converting it to the new CCNP type.
+	// This should not fail, so we are not returning any errors, just logging
+	// a warning.
```
Thanks for clarifying this. I meant to ask the previous time, but why are we ignoring the errors here? I see that we think that it should not fail, but _if it does_, shouldn't we return an error while logging as well?

fristonio replied:
I just thought since the validation we are doing here does not lead to an error, we shouldn't throw an error in the process either. But yeah I do think what you are saying makes sense, should I return an error here instead?

fristonio replied:
I have the made the changes, let me know if it requires more polishing.

christarazi reviewed (commented):
LGTM; just one question below.

pchaigno commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +109,90 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+		"instead of acting as a truly empty endpoint selector allowing all traffic. To " +
+		"ensure that the policy behavior does not affect your workloads, consider adding " +
+		"another policy that allows traffic to/from world and cluster entities. For a more " +
+		"detailed discussion on the topic, see GitHub issue #12844"
```
When will we want to remove this warning? When 1.10 is out? Once 1.9 is the oldest supported version? It would be nice to add a comment such as `We can remove this once we reach 1.x` so we can grep for it later and don't forget to remove it.

fristonio replied:
This makes sense, I will add a comment regarding this.
One thing I would like to discuss in this context is the backporting scenario for the changes. Do we plan to backport these changes to previous releases? If not we can remove the warnings once the oldest supported version is 1.9

pchaigno replied:
Given the upgrade concern, I doubt we can backport this to 1.8/1.7. And unless some users are hitting this, it's probably not worth it anyway.

pchaigno reviewed (approved):
LGTM. Just one minor nit below.

joestringer commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +109,89 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+		return err
+	}
+
+	// Print the warninig only once per CCNP.
```
I'm a little concerned about how frequently this log message could be printed, surely it's not more useful than once per Cilium run? ie wrap with `sync.Once()`?

fristonio replied:
Yeah, I think it makes sense to print the warning only once to hint the users about the changes. I will update the pull request.

joestringer commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +109,89 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+	if resCCNP.Specs != nil {
+		for _, rule := range resCCNP.Specs {
+			if containsEmptyToFromEndpoint(rule) {
+				logger.Warning(warnEmptyToFromEndpointMessage)
```
Will this log the warning even if the user intended for it to only select endpoints and not other entities? That's how I at least thought it worked from the beginning :/ Notifying users who are expecting the new behaviour is just extra noise. Would this be more appropriate in a release note / upgrade guide note instead?

fristonio replied:
I think if we are not planning to backport the pull request then it makes sense to remove this extra noise and have this information in the upgrade guide/release note. But if we will be backporting the Pull requests then I think we should not consider that users expect this behavior and have plumbed their policies accordingly. The warning does make sense in such cases. 
 What are your thoughts on this @aanm ?

joestringer reviewed (commented):
Looks good overall, only thing I'm wondering about is the log message. The good thing is it tells existing users if their policy is changing. The bad thing is that it'll also give new users a random extra warning potentially once per CCNP parse which might not even be relevant to them.

At most, I think we should print the warning once. I'm on the fence about whether it should be there at all, though I think if it's there and a user hits some issue because of it, at least they would know that this could be why. The opposite side is this could quickly become a false positive warning where an intended policy causes this warning then users get confused when there's not really a problem.

aanm commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +109,89 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+func containsEmptyToFromEndpoint(rule *api.Rule) bool {
+	if len(rule.Ingress) > 0 {
+		for _, r := range rule.Ingress {
+			if len(r.FromEndpoints) > 0 {
```
Why isn't it considered empty if we have other `r.From*` but `r.FromEndpoints == 0`?

fristonio replied:
The only place where we do the translation from `LabelSelector` to `EndpointSelector` is for `To/FromEndpoints` and `To/FromRequires`. From the documentation of `FromRequires` 

```
FromRequires is a list of additional constraints which must be met in order for the selected endpoints to be reachable. These additional constraints do no by itself grant access privileges and must always be accompanied with at least one matching FromEndpoints.
```
So we essentially only need to check if `FromEndpoints` contains a wildcard right, or am I missing something here?

aanm replied:
Please add a comment in the code stating :point_up:

fristonio committed:
docs: add wildcard from endpoints clusterwide policy example

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

aanm reviewed (approved):
There's only the backport concern that we need to address

kkourt commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +110,97 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+	// We can remove the check for this warning once 1.9 is the oldest supported Cilium version.
+	warnWildcardToFromEndpointMessage = "It seems you have a CiliumClusterwideNetworkPolicy " +
+		"with a wildcard to/from endpoint selector. The behavior of this selector has been " +
+		"changed. The selector now only allows traffic to/from Cilium managed K8s endpoints, " +
```
```suggestion
		"has been changed from 1.9 onwards. The selector now only allows traffic to/from Cilium managed K8s endpoints, " +
```

nit: I think it would be useful to mention from what version behavior changed. Based on the comment above, I'm assuming this changed on 1.9? (feel free to change the phrasing, mine was just a quick suggestion)

fristonio replied:
We are still discussing the backport scenario for the Pull Request. I will update the comment if required once we decide on that. Thanks for pointing it out though.

aanm replied:
We have decided in the Cilium meeting that this PR is going to be backported to 1.7 and 1.8 to avoid having different implementations of the policy in different Cilium versions.

kkourt commented on pkg/k8s/apis/cilium.io/v2/validator/validator.go:
```diff
@@ -105,10 +110,97 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
+		"instead of acting as a truly empty endpoint selector allowing all traffic. To " +
+		"ensure that the policy behavior does not affect your workloads, consider adding " +
+		"another policy that allows traffic to/from world and cluster entities. For a more " +
+		"detailed discussion on the topic, see GitHub issue #12844"
```
nit:

```suggestion
		"detailed discussion on the topic, see https://github.com/cilium/cilium/issues/12844"
```

kkourt reviewed (approved):
Just a couple of suggestions for your consideration.
Thanks!
